### ISSUE TITLE
Stokhos:  Errors while compiling with ensemble reduction

### ISSUE THREAD

**Issue by kliegeois** [trilinos/Trilinos#7067]
## Bug Report
@trilinos/stokhos 

### Description
If the develop branch is configured and compiled with ensemble reduction, compiling errors occur:

```
[ 76%] Building CXX object packages/stokhos/src/CMakeFiles/stokhos_muelu_mp_8_openmp.dir/MueLu_BlackBoxPFactory_MP_Vector_8_OpenMP.cpp.o
In file included from /home/kim/dev/Trilinos_ER/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialQRDenseSolver.hpp:54:0,
                 from /home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/Teuchos_SerialQRDenseSolver_MP_Vector.hpp:49,
                 from /home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp:65,
                 from /home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp:49,
                 from /home/kim/dev/Trilinos_ER/TrilinosERB/packages/stokhos/src/Stokhos_Tpetra_ETI_Helpers_MP_Vector_8_OpenMP.hpp:43,
                 from /home/kim/dev/Trilinos_ER/TrilinosERB/packages/stokhos/src/MueLu_BlackBoxPFactory_MP_Vector_8_OpenMP.cpp:51:
/home/kim/dev/Trilinos_ER/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialDenseSolver.hpp: In instantiation of ‘int Teuchos::SerialDenseSolver<OrdinalType, ScalarType>::applyRefinement() [with OrdinalType = int; ScalarType = Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >]’:
/home/kim/dev/Trilinos_ER/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialDenseSolver.hpp:724:62:   required from ‘int Teuchos::SerialDenseSolver<OrdinalType, ScalarType>::solve() [with OrdinalType = int; ScalarType = Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >]’
/home/kim/dev/Trilinos_ER/Trilinos/packages/muelu/src/Transfers/BlackBox/MueLu_BlackBoxPFactory_def.hpp:1637:7:   required from ‘void MueLu::BlackBoxPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::ComputeLocalEntries(const Teuchos::RCP<const Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> >&, Teuchos::Array<T>, Teuchos::Array<T>, MueLu::BlackBoxPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::LO, Teuchos::Array<T>, Teuchos::Array<T>, MueLu::BlackBoxPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::LO, Teuchos::Array<T>, Teuchos::Array<GlobalSizeT>, Teuchos::Array<GlobalSizeT>, Teuchos::Array<T>, Teuchos::Array<bool>, Teuchos::Array<int>, std::__cxx11::string, std::__cxx11::string, Teuchos::Array<T>, MueLu::BlackBoxPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::LO, Teuchos::Array<GlobalSizeT>, Teuchos::SerialDenseMatrix<LocalOrdinal, Scalar>&, Teuchos::SerialDenseMatrix<LocalOrdinal, Scalar>&, Teuchos::SerialDenseMatrix<LocalOrdinal, Scalar>&, Teuchos::Array<T>&, Teuchos::Array<T>&) const [with Scalar = Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >; LocalOrdinal = int; GlobalOrdinal = int; Node = Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP>; MueLu::BlackBoxPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::LO = int; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/home/kim/dev/Trilinos_ER/TrilinosERB/packages/stokhos/src/MueLu_BlackBoxPFactory_MP_Vector_8_OpenMP.cpp:61:1:   required from here
/home/kim/dev/Trilinos_ER/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialDenseSolver.hpp:753:3: error: no matching function for call to ‘Teuchos::SerialDenseSolver<int, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> > >::GERFS(const char&, int&, int&, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >*&, int&, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >*&, int&, __gnu_cxx::__alloc_traits<std::allocator<int> >::value_type*, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >*, int, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >*, int, __gnu_cxx::__alloc_traits<std::allocator<double> >::value_type*, __gnu_cxx::__alloc_traits<std::allocator<double> >::value_type*, __gnu_cxx::__alloc_traits<std::allocator<Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> > > >::value_type*, __gnu_cxx::__alloc_traits<std::allocator<int> >::value_type*, int*)’
   this->GERFS(ETranspChar[TRANS_], N_, NRHS, A_, LDA_, AF_, LDAF_, &IPIV_[0],
   ^~~~
In file included from /home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp:67:0,
                 from /home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp:49,
                 from /home/kim/dev/Trilinos_ER/TrilinosERB/packages/stokhos/src/Stokhos_Tpetra_ETI_Helpers_MP_Vector_8_OpenMP.hpp:43,
                 from /home/kim/dev/Trilinos_ER/TrilinosERB/packages/stokhos/src/MueLu_BlackBoxPFactory_MP_Vector_8_OpenMP.cpp:51:
/home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/Teuchos_LAPACK_MP_Vector.hpp:274:10: note: candidate: void Teuchos::LAPACK<OrdinalType, Sacado::MP::Vector<S> >::GERFS(char, OrdinalType, OrdinalType, const ScalarType*, OrdinalType, const ScalarType*, OrdinalType, const OrdinalType*, const ScalarType*, OrdinalType, Teuchos::LAPACK<OrdinalType, Sacado::MP::Vector<S> >::ScalarType*, OrdinalType, Teuchos::LAPACK<OrdinalType, Sacado::MP::Vector<S> >::ScalarType*, Teuchos::LAPACK<OrdinalType, Sacado::MP::Vector<S> >::ScalarType*, Teuchos::LAPACK<OrdinalType, Sacado::MP::Vector<S> >::ScalarType*, OrdinalType*, OrdinalType*) const [with OrdinalType = int; Storage = Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP>; Teuchos::LAPACK<OrdinalType, Sacado::MP::Vector<S> >::ScalarType = Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >]
     void GERFS(const char TRANS, const OrdinalType n, const OrdinalType nrhs, const ScalarType* A, const OrdinalType lda, const ScalarType* AF, const OrdinalType ldaf, const OrdinalType* IPIV, const ScalarType* B, const OrdinalType ldb, ScalarType* X, const OrdinalType ldx, ScalarType* FERR, ScalarType* BERR, ScalarType* WORK, OrdinalType* IWORK, OrdinalType* info) const
          ^~~~~
/home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/Teuchos_LAPACK_MP_Vector.hpp:274:10: note:   no known conversion for argument 13 from ‘__gnu_cxx::__alloc_traits<std::allocator<double> >::value_type* {aka double*}’ to ‘Teuchos::LAPACK<int, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> > >::ScalarType* {aka Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >*}’
In file included from /home/kim/dev/Trilinos_ER/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialQRDenseSolver.hpp:54:0,
                 from /home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/Teuchos_SerialQRDenseSolver_MP_Vector.hpp:49,
                 from /home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp:65,
                 from /home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp:49,
                 from /home/kim/dev/Trilinos_ER/TrilinosERB/packages/stokhos/src/Stokhos_Tpetra_ETI_Helpers_MP_Vector_8_OpenMP.hpp:43,
                 from /home/kim/dev/Trilinos_ER/TrilinosERB/packages/stokhos/src/MueLu_BlackBoxPFactory_MP_Vector_8_OpenMP.cpp:51:
/home/kim/dev/Trilinos_ER/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialDenseSolver.hpp: In instantiation of ‘int Teuchos::SerialDenseSolver<OrdinalType, ScalarType>::computeEquilibrateScaling() [with OrdinalType = int; ScalarType = Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >]’:
/home/kim/dev/Trilinos_ER/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialDenseSolver.hpp:838:53:   required from ‘int Teuchos::SerialDenseSolver<OrdinalType, ScalarType>::equilibrateRHS() [with OrdinalType = int; ScalarType = Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >]’
/home/kim/dev/Trilinos_ER/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialDenseSolver.hpp:660:26:   required from ‘int Teuchos::SerialDenseSolver<OrdinalType, ScalarType>::solve() [with OrdinalType = int; ScalarType = Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >]’
/home/kim/dev/Trilinos_ER/Trilinos/packages/muelu/src/Transfers/BlackBox/MueLu_BlackBoxPFactory_def.hpp:1637:7:   required from ‘void MueLu::BlackBoxPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::ComputeLocalEntries(const Teuchos::RCP<const Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> >&, Teuchos::Array<T>, Teuchos::Array<T>, MueLu::BlackBoxPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::LO, Teuchos::Array<T>, Teuchos::Array<T>, MueLu::BlackBoxPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::LO, Teuchos::Array<T>, Teuchos::Array<GlobalSizeT>, Teuchos::Array<GlobalSizeT>, Teuchos::Array<T>, Teuchos::Array<bool>, Teuchos::Array<int>, std::__cxx11::string, std::__cxx11::string, Teuchos::Array<T>, MueLu::BlackBoxPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::LO, Teuchos::Array<GlobalSizeT>, Teuchos::SerialDenseMatrix<LocalOrdinal, Scalar>&, Teuchos::SerialDenseMatrix<LocalOrdinal, Scalar>&, Teuchos::SerialDenseMatrix<LocalOrdinal, Scalar>&, Teuchos::Array<T>&, Teuchos::Array<T>&) const [with Scalar = Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >; LocalOrdinal = int; GlobalOrdinal = int; Node = Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP>; MueLu::BlackBoxPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::LO = int; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/home/kim/dev/Trilinos_ER/TrilinosERB/packages/stokhos/src/MueLu_BlackBoxPFactory_MP_Vector_8_OpenMP.cpp:61:1:   required from here
/home/kim/dev/Trilinos_ER/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialDenseSolver.hpp:776:3: error: no matching function for call to ‘Teuchos::SerialDenseSolver<int, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> > >::GEEQU(int&, int&, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >*&, int&, __gnu_cxx::__alloc_traits<std::allocator<double> >::value_type*, __gnu_cxx::__alloc_traits<std::allocator<double> >::value_type*, Teuchos::SerialDenseSolver<int, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> > >::MagnitudeType*, Teuchos::SerialDenseSolver<int, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> > >::MagnitudeType*, Teuchos::SerialDenseSolver<int, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> > >::MagnitudeType*, int*)’
   this->GEEQU (M_, N_, AF_, LDAF_, &R_[0], &C_[0], &ROWCND_, &COLCND_, &AMAX_, &INFO_);
   ^~~~
In file included from /home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp:67:0,
                 from /home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp:49,
                 from /home/kim/dev/Trilinos_ER/TrilinosERB/packages/stokhos/src/Stokhos_Tpetra_ETI_Helpers_MP_Vector_8_OpenMP.hpp:43,
                 from /home/kim/dev/Trilinos_ER/TrilinosERB/packages/stokhos/src/MueLu_BlackBoxPFactory_MP_Vector_8_OpenMP.cpp:51:
/home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/Teuchos_LAPACK_MP_Vector.hpp:270:10: note: candidate: void Teuchos::LAPACK<OrdinalType, Sacado::MP::Vector<S> >::GEEQU(OrdinalType, OrdinalType, const ScalarType*, OrdinalType, Teuchos::LAPACK<OrdinalType, Sacado::MP::Vector<S> >::ScalarType*, Teuchos::LAPACK<OrdinalType, Sacado::MP::Vector<S> >::ScalarType*, Teuchos::LAPACK<OrdinalType, Sacado::MP::Vector<S> >::ScalarType*, Teuchos::LAPACK<OrdinalType, Sacado::MP::Vector<S> >::ScalarType*, Teuchos::LAPACK<OrdinalType, Sacado::MP::Vector<S> >::ScalarType*, OrdinalType*) const [with OrdinalType = int; Storage = Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP>; Teuchos::LAPACK<OrdinalType, Sacado::MP::Vector<S> >::ScalarType = Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >]
     void GEEQU(const OrdinalType m, const OrdinalType n, const ScalarType* A, const OrdinalType lda, ScalarType* R, ScalarType* C, ScalarType* rowcond, ScalarType* colcond, ScalarType* amax, OrdinalType* info) const
          ^~~~~
/home/kim/dev/Trilinos_ER/Trilinos/packages/stokhos/src/sacado/kokkos/vector/Teuchos_LAPACK_MP_Vector.hpp:270:10: note:   no known conversion for argument 5 from ‘__gnu_cxx::__alloc_traits<std::allocator<double> >::value_type* {aka double*}’ to ‘Teuchos::LAPACK<int, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> > >::ScalarType* {aka Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 8, Kokkos::OpenMP> >*}’
packages/stokhos/src/CMakeFiles/stokhos_muelu_mp_8_openmp.dir/build.make:254: recipe for target 'packages/stokhos/src/CMakeFiles/stokhos_muelu_mp_8_openmp.dir/MueLu_BlackBoxPFactory_MP_Vector_8_OpenMP.cpp.o' failed
make[2]: *** [packages/stokhos/src/CMakeFiles/stokhos_muelu_mp_8_openmp.dir/MueLu_BlackBoxPFactory_MP_Vector_8_OpenMP.cpp.o] Error 1
CMakeFiles/Makefile2:15868: recipe for target 'packages/stokhos/src/CMakeFiles/stokhos_muelu_mp_8_openmp.dir/all' failed
make[1]: *** [packages/stokhos/src/CMakeFiles/stokhos_muelu_mp_8_openmp.dir/all] Error 2
Makefile:162: recipe for target 'all' failed
make: *** [all] Error 2
```

@etphipp, I think it could be great if one of the PR automatic test build can be configured with ensemble reduction set to on. It would be a way to ensure that the develop branch can always be compiled with and without ensemble reduction.

### Steps to Reproduce
1. Configure script: [config.txt](https://github.com/trilinos/Trilinos/files/4365203/config.txt)

**Comment by etphipp** [comment]
Thanks Kim!  The changes in #7068 look good to me, but I would like @trilinos/framework to comment on the PR testing change.  I worry that build will go away soon and this change will get lost.

**Comment by kliegeois** [comment]
@etphipp wrote:
> Thanks Kim! The changes in #7068 look good to me, but I would like @trilinos/framework to comment on the PR testing change. I worry that build will go away soon and this change will get lost.

I understand your worries. The version of the compiler for the build used to test the ensemble reduction has no particular importance for me providing that at least one build tests it. We can definitely use another build not to lose the modification.

**Comment by kliegeois** [comment]
The issue is closed by #7068.
Thanks for your inputs @etphipp !
